engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/loading/ShiftsLoadingTest.kt

```diff
@@ -0,0 +1,185 @@
+import info.metadude.android.eventfahrplan.engelsystem.models.ShiftsResult
+import info.metadude.kotlin.library.engelsystem.EngelsystemService
+import info.metadude.kotlin.library.engelsystem.models.Shift
+import kotlinx.coroutines.runBlocking
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import retrofit2.Call
+import retrofit2.Response
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+import java.io.EOFException
+import java.net.HttpURLConnection
+
+class ShiftsLoadingTest {
+
+    companion object {
+        private const val VALID_ONE_ITEM_SHIFTS_JSON = """
+                [
+                    {
+                        "Comment": "This is a very secret comment.",
+                        "Name": "Kirmes",
+                        "RID": 12,
+                        "SID": 579,
+                        "TID": 10,
+                        "UID": 32,
+                        "created_at_timestamp": 1565979486,
+                        "created_by_user_id": 32,
+                        "description": "Kirmes are fun.",
+                        "edited_at_timestamp": 1567367095,
+                        "edited_by_user_id": 32,
+                        "freeloaded": 0,
+                        "id": 37,
+                        "name": "Collect stickers",
+                        "shifttype_id": 6,
+                        "title": "Tag 1: Decorate fridge"
+                    }
+                ]
+                """
+        private const val INVALID_ONE_ITEM_SHIFTS_JSON = """
+                {
+                    "unknown": "foobar"
+                }
+                """
+        private const val EMPTY_ARRAY_SHIFTS_JSON = "[]"
+        private const val EMPTY_STRING = ""
+        private val ONE_ITEM_SHIFTS = listOf(Shift(
+                userComment = "This is a very secret comment.",
+                locationDescription = "Kirmes are fun.",
+                locationName = "Kirmes",
+                name = "Collect stickers",
+                sID = 579,
+                talkTitle = "Tag 1: Decorate fridge",
+                typeId = 6
+        ))
+    }
+
+    private val mockWebServer = MockWebServer()
+
+    @Before
+    fun setUp() {
+        mockWebServer.start()
+    }
+
+    @After
+    fun tearDown() {
+        mockWebServer.shutdown()
+    }
+
+    @Test
+    fun `Converts a success response with null body into an ShiftsResult exception`() {
+        val response: Response<List<Shift>> = Response.success(null)
+        val shiftsResult = response.toShiftsResult()
+        assertThat(shiftsResult).isEqualTo(ShiftsResult.Exception.MissingResponseSuccessBody)
+    }
+
+    @Test
+    fun `Converts a success response with empty body into an ShiftsResult success with an empty list`() {
+        val shiftsResult = Response.success(emptyList<Shift>()).toShiftsResult()
```

Comment:
I find it enormously helpful to separate tests into arrange, act, and assert sections. That way it's clear what you're testing exactly, i.e. which method. With a clear separation of inputs and expected results.

Here you're mixing the first two phases and I find it requires much more effort to read the test.

Reply:
:heavy_check_mark: Separated in d616cee9a6bd0f59e474a9a4e7c92efc45792b7a.